sound/soc/sof/intel/hda-stream.c

```diff
@@ -88,31 +88,46 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
 /*
  * set up Buffer Descriptor List (BDL) for host memory transfer
  * BDL describes the location of the individual buffers and is little endian.
+ *
+ * The typical BDL entries after set up should looks like this:
+ *
+ * bdl[0] from the init_offset,  n entries before wrap,
+ * the e-1 full entries, and the last e entry.
+ *
+ *			     init_offset
+ *				 |
+ * |------|...........|---------S-------|------|........|--------|-------|
+ *  bdl[n] bdl[n+j]    bdl[e-1] | bdl[0] bdl[1] bdl[m]   bdl[n-2] bdl[n-1]
```

Comment:
@keyonjie this is so difficult to follow. If I understand it correctly there are only 2 cases:
if init_offset ==0,there are bufsize/periodbytes (+1 in case it is not evenly divisible) periods
else there are 2 periods (or entries as you call it) right?

Reply:
@ranj063 the 2nd case what you mentioned can be more complicated, there can be > 2 periods there.
What the comment I wrote here is aim to abstract all possible cases to this common one. 

You can ignore n, m, j, e here, though they are useful when writing code in logic that Pierre mentioned:

1. from init_offset to first period start: bdl[0]
2. then all regular periods, bdl[m]
3. then end of buffer alignment, bdl[n-1]  ---- We have n items before buffer end;
4. then wrap until last period, bdl[e-2]
5. and again remaining bytes, bdl[e-1]. ---- We have e items in total.

Reply:
@keyonjie Yes, I understand the logic and ther is not doubt it is correct.

If I may make a suggestion: can we treat the case of a non-zero init_offset as 2 separate buffers,
then we end up with just as many bld entries except that the partial segments will be only at the end of the 2 buffer areas. For example: 
Case 1: init_offset = 0
init_offset
num of period_bytes size bdl entries = 5
num of partial bdl entries = 1
|
|------|------|------|------|------|---|

Case 2: non-zero init_offset
num of period_bytes size bdl entries = 4
num of partial bdl entries = 2
                 init_offset
                      |
|------|------|----|------|------|-----|

So the partial fragments are at the end of the 2 areas ( 0 - init_offset -1, init_offset - bufsize).

But with the current case there will be 3 partial entries. 

Anyway, this is only to improve maintainability. I am fine if the code is merged as is too.

Reply:
> @keyonjie Yes, I understand the logic and ther is not doubt it is correct.
> 
> If I may make a suggestion: can we treat the case of a non-zero init_offset as 2 separate buffers,
> then we end up with just as many bld entries except that the partial segments will be only at the end of the 2 buffer areas. For example:
> Case 1: init_offset = 0
> num of period_bytes size bdl entries = 5
> num of partial bdl entries = 1
> 
> init_offset = 0
> |
> |------|------|------|------|------|---|
> 
> Case 2: init_offset = 16
> num of period_bytes size bdl entries = 4
> num of partial bdl entries = 2
> 
> init_offset = 16
> 
> |------|------|----|------|------|-----|
> 
> So the partial fragments are at the end of the 2 areas ( 0 - init_offset -1, init_offset - bufsize).
> 
> But with the current case there will be 3 partial entries.

Sorry I don't catch you here, there will be only 2 partial entries in your sample.

> 
> Anyway, this is only to improve maintainability. I am fine if the code is merged as is too.

Reply:
@keyonjie what I meant was just break up the stream->bufsize into 2 and then assign bdl entries individually for [init_offset, buf_size) and then [0, init_offset -1). This way the logic will be a lot easier to follow. Let me take a swipe at it today and send you a patch

Reply:
@ranj063 the init_offset may be used for normal audio stream also, so we still need assign bdl entries for periods when init_offset!=0.